docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+      and used in CI/CD pipelines, without requiring users to install
+      additional tools such as Docker Engine or
+      [crane](https://github.com/google/go-containerregistry/blob/main/cmd/crane/README.md)
+      (or even ko, depending on how the builder is implemented). This eases the
+      path to adoption of Skaffold and reduces friction for users, both for
+      local development, and for anyone using Skaffold in CI/CD pipelines.
+
+  2.  The ko builder doesn't require running custom shell scripts. This means
+      more standardized builds, a desirable trait for enterprise users.
+
+The ko builder supports and enhances these Skaffold
+[features](https://skaffold.dev/docs/):
+
+- _fast local workflow_: building with ko is
+  [fast](https://cloud.google.com/blog/topics/developers-practitioners/ship-your-go-applications-faster-cloud-run-ko).
+
+- _share with other developers_: no additional tools are required to
+  `skaffold run` with the ko builder, not even Docker. Though if we don't embed
+  ko in Skaffold, ko will be a tool that all developers in a team would have to
+  install.
+
+- works great as a _CI/CD building block_: when using the ko builder, pipeline
+  steps can run using the default Skaffold container image, without
+  installing additional tools or keeping toolchain versions in sync across
+  local development and CI/CD.
+
+## Design
+
+Adding the ko builder requires making config changes to the Skaffold schema.
+
+1.  Add an entry to BuilderType enum in `proto/enums/enums.proto`:
+
+    ```proto
+        // ko Builder
+        KO = 7;
+    ```
+
+2.  Add a `KoArtifact` type:
+
+    ```go
+    type KoArtifact struct {
+    	// Annotations are key-value string pairs to add to the image manifest.
+    	// Also known as `LABEL` in `Dockerfile`s.
+    	// Ref: https://github.com/opencontainers/image-spec/blob/master/annotations.md
+    	Annotations map[string]string `yaml:"annotations,omitempty"`
+
+    	// BaseImage overrides the default ko base image.
+    	// Corresponds to, and overrides, the `defaultBaseImage` in `.ko.yaml`.
+    	BaseImage string `yaml:"fromImage,omitempty"`
+
+    	// Env are environment variables, in the `key=value` form, passed to the build.
+    	// For example: `CGO_ENABLED=1`.
+    	Env []string `yaml:"env,omitempty"`
+
+    	// Platforms is the list of platforms to build images for. Each platform
+    	// is of the format `os/arch[/variant]`, e.g., `linux/amd64`.
+    	// By default, the ko builder builds for `all` platforms supported by the
+    	// base image.
+    	Platforms []string `yaml:"platforms,omitempty"`
+
+    	// SourceDateEpoch is the `created` time of the container image.
+    	// Specify as the number of seconds since January 1st 1970, 00:00 UTC.
+    	// You can override this value by setting the `SOURCE_DATE_EPOCH`
+    	// environment variable.
+    	SourceDateEpoch unit64 `yaml:"sourceDateEpoch,omitempty"`
```

Comment:
would most users want this to be `time.Now`?

Reply:
I'd expect most would leave it at `0`, for reproducible builds. That's the ko default.

The proposal doesn't actually touch on reproducible builds yet, but we could elaborate on this. ko makes reproducible builds possible, for users who care about that. It requires a reproducible build environment though (consistent paths, `GOROOT`, etc) - until we add the ability to pass the `go build` flags required to remove paths from the binary.

Reply:
Just a nit: we can drop this field from config and rely on plumbing through `SOURCE_DATE_EPOCH` env variable for reproducible builds to the build process.
We can set it to `time.Now` when the env variable is not set.

Reply:
I'd suggest leaving the default to `0`. This is the ko default, and I see both Knative and Tekton uses this. It'd be confusing if ko behaved differently when used via Skaffold.

Your point of plumbing through envvars is good. Is there an established pattern in Skaffold for whether config in `skaffold.yaml takes precedence over external builder config (e.g., Jib config in `pom.xml`/`build.gradle`)?

Reply:
sounds good.